File: numberofislands.cpp
#include <vector>
#include <set>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <string>
using namespace std;

class Solution {
private:
    void bfs(vector<vector<char> >& grid, int m, int n, int i, int j) {
        queue<pair<int, int>> que;
        que.push(pair<int,int>{i, j});
        
        while(!que.empty()) {
            pair<int, int> indices = que.front();
            int i_ = indices.first;
            int j_ = indices.second;
            que.pop();
            if (i_ < 0 || j_ < 0 || i_ >= m || j_ >= n) {
                continue;
            }
            if (grid.at(i_).at(j_) == '1') {
                grid.at(i_).at(j_) = '2';
                que.push(pair<int,int>{i_+1, j_});
                que.push(pair<int,int>{i_, j_+1});
                que.push(pair<int,int>{i_, j_-1});
                que.push(pair<int,int>{i_-1, j_});
            }
        }
    }
public:
    int numIslands(vector<vector<char> >& grid) {
        int ans = 0;
        int m = grid.size();
        int n = grid.at(0).size();

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                char c = grid.at(i).at(j);
                if (c == '1') {
                    bfs(grid, m, n, i, j);
                    ans++;
                }
            }
        }
        return ans;
    }
};

File: numberOfIslands.py

from collections import deque
import queue
class Solution:
    def dfs(self, grid, i, j):
        if i >= 0 and i < self.m and j >= 0 and j < self.n and grid[i][j] == "1":
            grid[i][j] = "2"
            self.dfs(grid, i-1, j)
            self.dfs(grid, i+1, j)
            self.dfs(grid, i, j-1)
            self.dfs(grid, i, j+1)
        else:
            return

    def numIslands(self, grid: List[List[str]]) -> int:
        self.m = len(grid)
        self.n = len(grid[0])
        ans = 0
        que = queue.Queue()
        for i in range(self.m):
            for j in range(self.n):
                if grid[i][j] == "1":
                    ans += 1
                    grid[i][j] = "2"
                    que.put((i-1, j))
                    que.put((i+1, j))
                    que.put((i, j-1))
                    que.put((i, j+1))
                    while que.qsize() > 0:
                        y, x = que.get()
                        if 0 <= y < self.m and 0 <= x < self.n and grid[y][x] == "1":
                            grid[y][x] = "2"
                            que.put((y-1, x))
                            que.put((y+1, x))
                            que.put((y, x-1))
                            que.put((y, x+1))
        return ans
